Library/Homebrew/os/mac/cache_store.rb

```diff
@@ -0,0 +1,207 @@
+  attr_accessor :name
+
+  # Instantiates new `DatabaseCache` object
+  #
+  # @param  [String] name
+  # @return [nil]
+  def initialize(name)
+    @name = name
+  end
+
+  # Memoized `DBM` database object with on-disk database located in the
+  # `HOMEBREW_CACHE`
+  #
+  # @return [DBM] db
+  def db
+    @db ||= DBM.open("#{HOMEBREW_CACHE}/#{name}", 0666, DBM::WRCREAT)
+  end
+
+  # Close the `DBM` database object after usage
+  #
+  # @return [nil]
+  def close
+    db.close
+  end
+end
+
+#
+# `CacheStore` is an abstract base class which provides methods to mutate and
+# fetch data from a persistent storage mechanism
+#
+# @abstract
+#
+class CacheStore
+  # Instantiates a new `CacheStore` class
+  #
+  # @param  [DatabaseCache] database_cache
+  # @return [nil]
+  def initialize(database_cache)
+    @db = database_cache.db
+  end
+
+  # Inserts new values or updates existing cached values to persistent storage
+  # mechanism
+  #
+  # @abstract
+  # @param  [Any]
+  # @return [nil]
+  def update!(*)
+    raise NotImplementedError
+  end
+
+  # Fetches cached values in persistent storage according to the type of data
+  # stored
+  #
+  # @abstract
+  # @param  [Any]
+  # @return [Any]
+  def fetch(*)
+    raise NotImplementedError
+  end
+
+  # Deletes data from the cache based on a condition defined in a concrete class
+  #
+  # @abstract
+  # @return [nil]
+  def flush_cache!
+    raise NotImplementedError
+  end
+
+  protected
+
+  # A class instance providing access to the `DBM` database object
+  #
+  # @return [DBM]
+  attr_reader :db
+end
+
+#
+# `LinkageStore` is a concrete class providing methods to fetch and mutate
+# linkage-specific data used by the `brew linkage` command
+#
+# If the cache hasn't changed, don't do extra processing in `LinkageChecker`.
+# Instead, just fetch the data stored in the cache
+#
+class LinkageStore < CacheStore
+  # Types of dylibs of the form (label -> array)
+  HASH_LINKAGE_TYPES = %w[brewed_dylibs reverse_links].freeze
+
+  # The keg name for the `LinkageChecker` class
+  #
+  # @return [String]
+  attr_reader :key
+
+  # Initializes new `LinkageStore` class
+  #
+  # @param  [String]        keg_name
+  # @param  [DatabaseCache] database_cache
+  # @return [nil]
+  def initialize(keg_name, database_cache)
+    @key = keg_name
+    super(database_cache)
+  end
+
+  # Inserts new values or updates existing cached values to persistent storage
+  # mechanism according to the type of data
+  #
+  # @param  [Hash] path_values
+  # @param  [Hash] hash_values
+  # @return [nil]
+  def update!(
+    path_values: {
+      "system_dylibs" => %w[], "variable_dylibs" => %w[], "broken_dylibs" => %w[],
+      "indirect_deps" => %w[], "undeclared_deps" => %w[], "unnecessary_deps" => %w[]
+    },
+    hash_values: {
+      "brewed_dylibs" => {}, "reverse_links" => {}
+    }
+  )
+    db[key] = {
+      "path_values" => format_path_values(path_values),
+      "hash_values" => format_hash_values(hash_values),
```

Comment:
What's the difference between a path and hash value?

Reply:
In `LinkageChecker::display_items`, the output can vary depending on if the type of `things` is an `Array` or `Hash`.

`hash_values` are values which store the `list_label` in addition to the list `items`, whereas `path_values` are values that just store the list `items`.

If you have suggestions for alternate ways to name the variables, I'd be happy to hear any suggestions. 🙂 

```ruby
  def display_items(label, things)
    return if things.empty?
    puts "#{label}:"
    if things.is_a? Hash
      things.sort.each do |list_label, list|
        list.sort.each do |item|
          puts "  #{item} (#{list_label})"
        end
      end
    else
      things.sort.each do |item|
        puts "  #{item}"
      end
    end
  end
```

Reply:
I understand the differentiation now, thanks, but not why e.g. `things.is_a? Hash` is sufficient and you need to store them separately?